File: ios/HMSHelper.swift
import HMSSDK
import Foundation

class HMSHelper: NSObject {
    static var audioMixerSourceHashMap: [String: HMSAudioNode]?

    static func getUnavailableRequiredKey(_ data: NSDictionary, _ requiredKeys: [String]) -> String {
        for (key) in requiredKeys {
            let value = data.object(forKey: key)
            if value == nil {
                return key + "_Is_Required"
            }
            if value is NSNull {
                return key + "_Is_Null"
            }
        }
        return "SEND_ALL_REQUIRED_KEYS"
    }

    static func getPeerFromPeerId(_ peerID: String?, remotePeers: [HMSRemotePeer]?, localPeer: HMSLocalPeer?) -> HMSPeer? {

        guard let peerID = peerID, let peers = remotePeers else { return nil }
        if peerID == localPeer?.peerID {
            return localPeer
        }
        return peers.first { $0.peerID == peerID }
    }

    static func getRemotePeerFromPeerId(_ peerID: String?, remotePeers: [HMSRemotePeer]?) -> HMSPeer? {

        guard let peerID = peerID, let peers = remotePeers else { return nil }

        return peers.first { $0.peerID == peerID }
    }

    static func getRolesFromRoleNames(_ targetedRoles: [String]?, roles: [HMSRole]?) -> [HMSRole] {

        guard let roles = roles,
              let targetedRoles = targetedRoles
        else { return [HMSRole]() }

        return roles.filter { targetedRoles.contains($0.name) }
    }

    static func getRoleFromRoleName(_ role: String?, roles: [HMSRole]?) -> HMSRole? {

        guard let roles = roles, let roleName = role else { return nil }

        return roles.first { $0.name == roleName }
    }

    static func getLocalTrackFromTrackId(_ trackID: String?, localPeer: HMSLocalPeer?) -> HMSTrack? {
        if localPeer?.videoTrack?.trackId == trackID {
            return localPeer?.videoTrack
        }

        if localPeer?.audioTrack?.trackId == trackID {
            return localPeer?.audioTrack
        }

        for track in localPeer?.auxiliaryTracks ?? [] where track.trackId == trackID {
            return track
        }

        return nil
    }

    static func getRemoteAudioTrackFromTrackId(_ trackID: String?, _ remotePeers: [HMSRemotePeer]?) -> HMSRemoteAudioTrack? {
        for peer in remotePeers ?? [] {
            if peer.audioTrack?.trackId == trackID {
                return peer.audioTrack as? HMSRemoteAudioTrack
            }
        }
        return nil
    }

    static func getRemoteAudioAuxiliaryTrackFromTrackId(_ trackID: String?, _ remotePeers: [HMSRemotePeer]?) -> HMSRemoteAudioTrack? {
        for peer in remotePeers ?? [] {
            if peer.audioTrack?.trackId == trackID {
                return peer.audioTrack as? HMSRemoteAudioTrack
            }
            let auxTracks = peer.auxiliaryTracks

            for track in auxTracks ?? [] {
                if track.kind == HMSTrackKind.audio && track.trackId == trackID {
                    return track as? HMSRemoteAudioTrack
                }
            }
        }
        return nil
    }

    static func getRemoteVideoTrackFromTrackId(_ trackID: String?, _ remotePeers: [HMSRemotePeer]?) -> HMSRemoteVideoTrack? {
        for peer in remotePeers ?? [] {
            if peer.videoTrack?.trackId == trackID {
                return peer.videoTrack as? HMSRemoteVideoTrack
            }
        }
        return nil
    }

    static func getHmsTrackType(_ kind: HMSTrackKind?) -> String? {
        if kind == HMSTrackKind.video {
            return "VIDEO"
        } else if kind == HMSTrackKind.audio {
            return "AUDIO"
        }
        return nil
    }

    static func getTrackFromTrackId(_ trackID: String?, _ remotePeers: [HMSRemotePeer]?) -> HMSTrack? {

        for peer in remotePeers ?? [] {
            if peer.videoTrack?.trackId == trackID {
                return peer.videoTrack
            }

            if peer.audioTrack?.trackId == trackID {
                return peer.audioTrack
            }

            for track in peer.auxiliaryTracks ?? [] where track.trackId == trackID {
                return track
            }
        }

        return nil
    }

    static func getHms(_ credentials: NSDictionary, _ hmsCollection: [String: HMSRNSDK]) -> HMSRNSDK? {
        guard let id = credentials.value(forKey: "id") as? String,
              let hms = hmsCollection[id]
        else {
            return nil
        }
        return hms
    }

    static func getFrameworkInfo(_ frameworkInfo: NSDictionary?) -> HMSFrameworkInfo? {
        guard let data = frameworkInfo,
              let version = data.value(forKey: "version") as? String,
              let sdkVersion = data.value(forKey: "sdkVersion") as? String
        else {
            return nil
        }
        return HMSFrameworkInfo(type: HMSFrameworkType.reactNative, version: version, sdkVersion: sdkVersion)
    }

    static func getLocalVideoSettings(_ settings: NSDictionary?) -> HMSVideoTrackSettings? {
        if settings === nil {
            return nil
        }
        let codec = HMSCodec.VP8
        let resolution = HMSVideoResolution.init(width: 320, height: 180)
        let maxBitrate = 512
        let maxFrameRate = 25
        let trackDescription = ""
        let cameraFacing = settings?.value(forKey: "cameraFacing") as? String
        let cameraFacingEncoded = HMSHelper.getCameraFacing(cameraFacing)
        let initialState = settings?.value(forKey: "initialState") as? String
        let initialStateEncoded = HMSHelper.getHMSTrackSettingsInitState(initialState)
        let hmsTrackSettings = HMSVideoTrackSettings(codec: codec,
                                                     resolution: resolution,
                                                     maxBitrate: maxBitrate,
                                                     maxFrameRate: maxFrameRate,
                                                     cameraFacing: cameraFacingEncoded,
                                                     simulcastSettings: nil,
                                                     trackDescription: trackDescription,
                                                     initialMuteState: initialStateEncoded,
                                                     videoPlugins: nil)
        return hmsTrackSettings
    }

    static func getLocalAudioSettings(_ settings: NSDictionary?, _ hms: HMSSDK?, _ delegate: HMSManager?, _ id: String) -> HMSAudioTrackSettings? {
        if settings === nil {
            return nil
        }
        let initialState = settings?.value(forKey: "initialState") as? String
        let initialStateEncoded = HMSHelper.getHMSTrackSettingsInitState(initialState)

        if #available(iOS 13.0, *) {
            var audioMixerSourceMap: [String: HMSAudioNode]?
            if let playerNode = settings?.value(forKey: "audioSource") as? [String] {
                audioMixerSourceMap = [String: HMSAudioNode]()
                for node in playerNode {
                    if audioMixerSourceMap?[node] == nil {
                        if node == "mic_node" {
                            audioMixerSourceMap?["mic_node"] = HMSMicNode()
                        } else if node == "screen_broadcast_audio_receiver_node" {
                            do {
                                audioMixerSourceMap?["screen_broadcast_audio_receiver_node"] = try hms?.screenBroadcastAudioReceiverNode()
                            } catch {
                                delegate?.emitEvent("ON_ERROR", ["error": ["code": 6002, "description": error.localizedDescription, "isTerminal": false, "canRetry": true, "params": ["function": #function]], "id": id])
                            }
                        } else {
                            audioMixerSourceMap?[node] = HMSAudioFilePlayerNode()
                        }
                    }
                }
            }

            if let audioMixerSourceMap = audioMixerSourceMap {
                do {
                    self.audioMixerSourceHashMap = audioMixerSourceMap
                    let audioMixerSource = try HMSAudioMixerSource(nodes: audioMixerSourceMap.values.map {$0})
                    return HMSAudioTrackSettings(maxBitrate: 32, trackDescription: "", initialMuteState: initialStateEncoded, audioSource: audioMixerSource)
                } catch {
                    delegate?.emitEvent("ON_ERROR", ["error": ["code": 6002, "description": error.localizedDescription, "isTerminal": false, "canRetry": true, "params": ["function": #function]], "id": id])
                    return nil
                }
            }
        }

        return HMSAudioTrackSettings(maxBitrate: 32, trackDescription: "", initialMuteState: initialStateEncoded, audioSource: nil)
    }

    static func getAudioMixerSourceMap() -> [String: HMSAudioNode]? {
        return self.audioMixerSourceHashMap
    }

    static func getVideoResolution(_ data: [String: Double]) -> HMSVideoResolution? {
        guard let width = data["width"],
              let height = data["height"]
        else {
            return nil
        }

        return HMSVideoResolution.init(width: width, height: height)
    }

    static func getVideoCodec(_ codecString: String?) -> HMSCodec {
        switch codecString {
        case "H264":
            return HMSCodec.H264
        case "VP8":
            return HMSCodec.VP8
        default:
            return HMSCodec.H264
        }
    }

    static func getCameraFacing(_ cameraFacing: String?) -> HMSCameraFacing {
        switch cameraFacing {
        case "FRONT":
            return HMSCameraFacing.front
        case "BACK":
            return HMSCameraFacing.back
        default:
            return HMSCameraFacing.front
        }
    }

    static func getHMSTrackSettingsInitState(_ initState: String?) -> HMSTrackMuteState {
        switch initState {
        case "MUTED":
            return HMSTrackMuteState.mute
        case "UNMUTED":
            return HMSTrackMuteState.unmute
        default:
            return HMSTrackMuteState.unmute
        }
    }

    static func getHMSTrackInitState(_ initState: HMSTrackMuteState?) -> String {
        switch initState {
        case .mute:
            return "MUTED"
        case .unmute:
            return "UNMUTED"
        default:
            return "UNMUTED"
        }
    }

    static func getRtmpUrls(_ strings: [String]?) -> [URL]? {
        if let extractedStrings = strings {
            var arr: [URL] = []
            for urlString in extractedStrings {
                let urlInstance = URL(string: urlString)
                if let urlExtracted = urlInstance {
                    arr.append(urlExtracted)
                }
            }
            return arr
        } else {
            return nil
        }
    }

    static func getHMSHLSMeetingURLVariants(_ variants: [[String: Any]]?) -> [HMSHLSMeetingURLVariant] {
        var hlsVariants: [HMSHLSMeetingURLVariant] = []
        for variant in variants ?? [] {
            let meetingURLVariant = HMSHelper.getHMSHLSMeetingURLVariant(variant)
            if let extractedVariant = meetingURLVariant {
                hlsVariants.append(extractedVariant)
            }
        }
        return hlsVariants
    }

    static func getHlsRecordingConfig(_ config: NSDictionary?) -> HMSHLSRecordingConfig? {
        guard let hlsRecordingConfig = config
        else {
            return nil
        }
        let singleFilePerLayer = hlsRecordingConfig.value(forKey: "singleFilePerLayer") as? Bool
        let videoOnDemand = hlsRecordingConfig.value(forKey: "videoOnDemand") as? Bool

        return HMSHLSRecordingConfig(singleFilePerLayer: singleFilePerLayer ?? false, enableVOD: videoOnDemand ?? false)
    }

    static func getHMSHLSMeetingURLVariant(_ variant: [String: Any]) -> HMSHLSMeetingURLVariant? {
        let meetingUrl = variant["meetingUrl"] as? String
        let metadata = variant["metadata"] as? String

        if let extractedUrl = meetingUrl, let url = URL(string: extractedUrl) {
            return HMSHLSMeetingURLVariant(meetingURL: url, metadata: metadata ?? "")
        }
        return nil
    }
}

File: packages/react-native-hms/android/src/main/java/com/reactnativehmssdk/HMSHelper.kt
package com.reactnativehmssdk

import android.content.Context
import android.graphics.Bitmap
import android.os.Build
import android.os.Handler
import android.util.Base64
import android.util.Log
import android.view.PixelCopy
import android.webkit.URLUtil
import androidx.annotation.RequiresApi
import com.facebook.react.bridge.*
import com.facebook.react.uimanager.events.RCTEventEmitter
import live.hms.video.audio.HMSAudioManager
import live.hms.video.error.HMSException
import live.hms.video.events.AgentType
import live.hms.video.media.settings.*
import live.hms.video.media.tracks.HMSRemoteAudioTrack
import live.hms.video.media.tracks.HMSRemoteVideoTrack
import live.hms.video.media.tracks.HMSTrack
import live.hms.video.sdk.HMSSDK
import live.hms.video.sdk.listeners.PeerListResultListener
import live.hms.video.sdk.models.*
import live.hms.video.sdk.models.enums.AudioMixingMode
import live.hms.video.sdk.models.role.HMSRole
import live.hms.video.services.LogAlarmManager
import live.hms.video.utils.HMSLogger
import live.hms.video.utils.HmsUtilities
import org.webrtc.SurfaceViewRenderer
import java.io.ByteArrayOutputStream
import java.util.*
import kotlin.collections.ArrayList
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

object HMSHelper {
  fun areAllRequiredKeysAvailable(
    map: ReadableMap?,
    requiredKeys: Array<Pair<String, String>>,
  ): Boolean {
    if (map == null) {
      return false
    }
    for ((key, value) in requiredKeys) {
      if (map.hasKey(key)) {
        when (value) {
          "String" -> {
            if (map.getString(key) == null) {
              return false
            }
          }
          "Array" -> {
            if (map.getArray(key) == null) {
              return false
            }
          }
          "Map" -> {
            if (map.getMap(key) == null) {
              return false
            }
          }
        }
      } else {
        return false
      }
    }
    return true
  }

  fun getUnavailableRequiredKey(
    map: ReadableMap?,
    requiredKeys: Array<Pair<String, String>>,
  ): String? {
    if (map == null) {
      return "Object_Is_Null"
    }
    for ((key, value) in requiredKeys) {
      if (map.hasKey(key)) {
        when (value) {
          "String" -> {
            if (map.getString(key) == null) {
              return key + "_Is_Null"
            }
          }
          "Array" -> {
            if (map.getArray(key) == null) {
              return key + "_Is_Null"
            }
          }
          "Map" -> {
            if (map.getMap(key) == null) {
              return key + "_Is_Null"
            }
          }
        }
      } else {
        return key + "_Is_Required"
      }
    }
    return null
  }

  fun getPeerFromPeerId(
    peerId: String?,
    room: HMSRoom?,
  ): HMSPeer? {
    if (peerId != null && room != null) {
      return HmsUtilities.getPeer(peerId, room)
    }
    return null
  }

  suspend fun getRemotePeerFromPeerId(
    peerId: String?,
    hmsSDK: HMSSDK?,
  ): HMSRemotePeer? {
    return suspendCoroutine {
      val room = hmsSDK?.getRoom()

      if (peerId != null && room != null) {
        val peerFromRoom = HmsUtilities.getPeer(peerId, room) as? HMSRemotePeer
        if (peerFromRoom != null) {
          it.resume(peerFromRoom)
        } else {
          val limit = 1
          val peerIds = arrayListOf(peerId)
          val peerListIterator = hmsSDK.getPeerListIterator(PeerListIteratorOptions(limit = limit, byPeerIds = peerIds))
          peerListIterator.next(object : PeerListResultListener {
            override fun onError(error: HMSException) {
              it.resumeWithException(error)
            }
            override fun onSuccess(result: ArrayList<HMSPeer>) {
              val peerFromIterator = result[0]
              it.resume(peerFromIterator as? HMSRemotePeer)
            }
          })
        }
      } else {
        it.resume(null)
      }
    }
  }

  fun getRolesFromRoleNames(
    targetedRoles: ArrayList<String>?,
    roles: List<HMSRole>?,
  ): List<HMSRole> {
    val encodedRoles: MutableList<HMSRole> = mutableListOf()

    if (targetedRoles != null && roles != null) {
      for (role in roles) {
        for (targetedRole in targetedRoles) {
          if (targetedRole == role.name) {
            encodedRoles.add(role)
          }
        }
      }
    }
    return encodedRoles.toList()
  }

  fun getRoleFromRoleName(
    role: String?,
    roles: List<HMSRole>?,
  ): HMSRole? {
    if (role != null && roles != null) {
      for (hmsRole in roles) {
        if (role == hmsRole.name) {
          return hmsRole
        }
      }
    }
    return null
  }

  fun getRemoteAudioTrackFromTrackId(
    trackId: String?,
    room: HMSRoom?,
  ): HMSRemoteAudioTrack? {
    if (trackId != null && room != null) {
      return HmsUtilities.getAudioTrack(trackId, room) as? HMSRemoteAudioTrack
    }
    return null
  }

  fun getRemoteVideoTrackFromTrackId(
    trackId: String?,
    room: HMSRoom?,
  ): HMSRemoteVideoTrack? {
    if (trackId != null && room != null) {
      return HmsUtilities.getVideoTrack(trackId, room) as? HMSRemoteVideoTrack
    }
    return null
  }

  fun getTrackFromTrackId(
    trackId: String?,
    room: HMSRoom?,
  ): HMSTrack? {
    if (trackId != null && room != null) {
      return HmsUtilities.getTrack(trackId, room)
    }
    return null
  }

  fun getFrameworkInfo(data: ReadableMap?): FrameworkInfo? {
    if (data != null &&
      this.areAllRequiredKeysAvailable(
        data,
        arrayOf(Pair("version", "String"), Pair("sdkVersion", "String")),
      )
    ) {
      val version = data.getString("version") as String
      val sdkVersion = data.getString("sdkVersion") as String
      val isPrebuilt = data.getBoolean("isPrebuilt")
      return FrameworkInfo(AgentType.REACT_NATIVE, sdkVersion, version, isPrebuilt)
    }
    return null
  }

  fun getLogSettings(data: ReadableMap?): HMSLogSettings? {
    if (data != null &&
      this.areAllRequiredKeysAvailable(
        data,
        arrayOf(
          Pair("level", "String"),
          Pair("maxDirSizeInBytes", "String"),
          Pair("isLogStorageEnabled", "Boolean"),
        ),
      )
    ) {
      val level = getLogLevel(data.getString("level"))
      val maxDirSizeInBytes = getLogAlarmManager(data.getString("maxDirSizeInBytes"))
      val isLogStorageEnabled = data.getBoolean("isLogStorageEnabled")
      return HMSLogSettings(maxDirSizeInBytes, isLogStorageEnabled, level)
    }
    return null
  }

  private fun getLogAlarmManager(logAlarmManager: String?): Long {
    when (logAlarmManager) {
      "DEFAULT_DIR_SIZE" -> {
        return LogAlarmManager.DEFAULT_DIR_SIZE
      }
      "DEFAULT_LOGS_FILE_NAME" -> {
        return LogAlarmManager.DEFAULT_LOGS_FILE_NAME.toLong()
      }
      "MAX_DIR_SIZE" -> {
        return LogAlarmManager.MAX_DIR_SIZE.toLong()
      }
    }
    return LogAlarmManager.DEFAULT_DIR_SIZE
  }

  private fun getLogLevel(logLevel: String?): HMSLogger.LogLevel {
    when (logLevel) {
      "VERBOSE" -> {
        return HMSLogger.LogLevel.VERBOSE
      }
      "WARN" -> {
        return HMSLogger.LogLevel.WARN
      }
      "ERROR" -> {
        return HMSLogger.LogLevel.ERROR
      }
    }
    return HMSLogger.LogLevel.VERBOSE
  }

  fun getTrackSettings(data: ReadableMap?): HMSTrackSettings? {
    if (data == null) {
      return null
    }
    val builder = HMSTrackSettings.Builder()
    if (this.areAllRequiredKeysAvailable(data, arrayOf(Pair("video", "Map")))) {
      val video = data.getMap("video")
      val videoSettings = this.getVideoTrackSettings(video)
      if (videoSettings != null) {
        builder.video(videoSettings)
      }
    }

    if (this.areAllRequiredKeysAvailable(data, arrayOf(Pair("audio", "Map")))) {
      val audio = data.getMap("audio")
      val audioSettings = this.getAudioTrackSettings(audio)
      if (audioSettings != null) {
        builder.audio(audioSettings)
      }
    }
    return builder.build()
  }

  private fun getAudioTrackSettings(data: ReadableMap?): HMSAudioTrackSettings? {
    if (data == null) {
      return null
    }

    val builder = HMSAudioTrackSettings.Builder()
    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("useHardwareEchoCancellation", "Boolean")))
    ) {
      val useHardwareEchoCancellation = data.getBoolean("useHardwareEchoCancellation")
      builder.setUseHardwareAcousticEchoCanceler(useHardwareEchoCancellation)
    }

    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("initialState", "String")))) {
      val initialState = getHMSTrackSettingsInitState(data.getString("initialState"))
      builder.initialState(initialState)
    }
    return builder.build()
  }

  private fun getVideoTrackSettings(data: ReadableMap?): HMSVideoTrackSettings? {
    if (data == null) {
      return null
    }
    val builder = HMSVideoTrackSettings.Builder()

    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("cameraFacing", "String")))) {
      val cameraFacing = getCameraFacing(data.getString("cameraFacing"))
      builder.cameraFacing(cameraFacing)
    }

    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("disableAutoResize", "Boolean")))) {
      val disableAutoResize = data.getBoolean("disableAutoResize")
      builder.disableAutoResize(disableAutoResize)
    }

    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("initialState", "String")))) {
      val initialState = getHMSTrackSettingsInitState(data.getString("initialState"))
      builder.initialState(initialState)
    }

    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("forceSoftwareDecoder", "Boolean")))) {
      val forceSoftwareDecoder = data.getBoolean("forceSoftwareDecoder")
      builder.forceSoftwareDecoder(forceSoftwareDecoder)
    }
    return builder.build()
  }

  private fun getHMSTrackSettingsInitState(initState: String?): HMSTrackSettings.InitState {
    when (initState) {
      "UNMUTED" -> {
        return HMSTrackSettings.InitState.UNMUTED
      }
      "MUTED" -> {
        return HMSTrackSettings.InitState.MUTED
      }
    }
    return HMSTrackSettings.InitState.UNMUTED
  }

  fun getAudioMixingMode(audioMixingMode: String?): AudioMixingMode {
    when (audioMixingMode) {
      "TALK_ONLY" -> {
        return AudioMixingMode.TALK_ONLY
      }
      "MUSIC_ONLY" -> {
        return AudioMixingMode.MUSIC_ONLY
      }
      "TALK_AND_MUSIC" -> {
        return AudioMixingMode.TALK_AND_MUSIC
      }
    }
    return AudioMixingMode.TALK_AND_MUSIC
  }

  private fun getCameraFacing(cameraFacing: String?): HMSVideoTrackSettings.CameraFacing {
    when (cameraFacing) {
      "FRONT" -> {
        return HMSVideoTrackSettings.CameraFacing.FRONT
      }
      "BACK" -> {
        return HMSVideoTrackSettings.CameraFacing.BACK
      }
    }
    return HMSVideoTrackSettings.CameraFacing.FRONT
  }

  fun getHms(
    credentials: ReadableMap,
    hmsCollection: MutableMap<String, HMSRNSDK>,
  ): HMSRNSDK? {
    val id = credentials.getString("id")

    return if (id != null) {
      hmsCollection[id]
    } else {
      null
    }
  }

  private fun getRtmpUrls(rtmpURLsList: ReadableArray?): List<String> {
    val rtmpURLs = mutableListOf<String>()
    if (rtmpURLsList !== null) {
      for (rtmpURL in rtmpURLsList.toArrayList()) {
        rtmpURLs.add(rtmpURL as String)
      }
    }
    return rtmpURLs
  }

  fun getHLSConfig(data: ReadableMap?): HMSHLSConfig? {
    if (data === null) {
      return data
    }
    var hlsMeetingUrlVariant: List<HMSHLSMeetingURLVariant>? = null
    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("meetingURLVariants", "Array")))) {
      val meetingURLVariants =
        data.getArray("meetingURLVariants")?.toArrayList() as? ArrayList<HashMap<String, String>>
      hlsMeetingUrlVariant = getHMSHLSMeetingURLVariants(meetingURLVariants)
    }
    var hlsRecordingConfig: HMSHlsRecordingConfig? = null
    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("hlsRecordingConfig", "Map")))) {
      val recordingConfig = data.getMap("hlsRecordingConfig")
      hlsRecordingConfig = getHlsRecordingConfig(recordingConfig)
    }
    return HMSHLSConfig(hlsMeetingUrlVariant, hlsRecordingConfig)
  }

  private fun getHMSHLSMeetingURLVariants(hmsMeetingURLVariants: ArrayList<HashMap<String, String>>?): List<HMSHLSMeetingURLVariant> {
    val meetingURLVariants = mutableListOf<HMSHLSMeetingURLVariant>()
    if (hmsMeetingURLVariants !== null) {
      for (variant in hmsMeetingURLVariants) {
        val meetingURLVariant = this.getHMSHLSMeetingURLVariant(variant)
        meetingURLVariants.add(meetingURLVariant)
      }
    }
    return meetingURLVariants
  }

  private fun getHlsRecordingConfig(hmsHlsRecordingConfig: ReadableMap?): HMSHlsRecordingConfig? {
    if (hmsHlsRecordingConfig !== null) {
      var singleFilePerLayer = false
      var videoOnDemand = false
      if (areAllRequiredKeysAvailable(
          hmsHlsRecordingConfig,
          arrayOf(Pair("singleFilePerLayer", "Boolean")),
        )
      ) {
        singleFilePerLayer = hmsHlsRecordingConfig.getBoolean("singleFilePerLayer")
      }
      if (areAllRequiredKeysAvailable(
          hmsHlsRecordingConfig,
          arrayOf(Pair("videoOnDemand", "Boolean")),
        )
      ) {
        videoOnDemand = hmsHlsRecordingConfig.getBoolean("videoOnDemand")
      }
      return HMSHlsRecordingConfig(singleFilePerLayer, videoOnDemand)
    }
    return null
  }

  private fun getHMSHLSMeetingURLVariant(hmsMeetingURLVariant: HashMap<String, String>?): HMSHLSMeetingURLVariant {
    var meetingURLVariant = HMSHLSMeetingURLVariant("", "")
    if (hmsMeetingURLVariant !== null) {
      val meetingUrl = hmsMeetingURLVariant["meetingUrl"] as String
      val metadata = hmsMeetingURLVariant["metadata"] as String
      meetingURLVariant = HMSHLSMeetingURLVariant(meetingUrl, metadata)
    }
    return meetingURLVariant
  }

  private fun getResolution(data: ReadableMap): HMSRtmpVideoResolution {
    val height = data.getInt("height")
    val width = data.getInt("width")
    return HMSRtmpVideoResolution(width, height)
  }

  fun getRtmpConfig(data: ReadableMap): HMSRecordingConfig? {
    val record = data.getBoolean("record")
    var meetingURL: String? = null
    var rtmpURLs = listOf<String>()
    var resolution: HMSRtmpVideoResolution? = null
    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("meetingURL", "String")))) {
      val meetingURLValid = data.getString("meetingURL") as String

      if (meetingURLValid.isNotEmpty()) {
        if (URLUtil.isValidUrl(meetingURLValid)) {
          meetingURL = meetingURLValid
        } else {
          return null
        }
      }
    }

    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("rtmpURLs", "Array")))) {
      val rtmpURLsValid = data.getArray("rtmpURLs")
      rtmpURLs = this.getRtmpUrls(rtmpURLsValid)
    }

    if (areAllRequiredKeysAvailable(data, arrayOf(Pair("resolution", "Map")))) {
      val resolutionValid = data.getMap("resolution")
      resolution = resolutionValid?.let { this.getResolution(it) }
    }
    return HMSRecordingConfig(meetingURL, rtmpURLs, record, resolution)
  }

  fun getHmsConfig(credentials: ReadableMap): HMSConfig {
    var config =
      HMSConfig(
        credentials.getString("username") as String,
        credentials.getString("authToken") as String,
      )

    when {
      areAllRequiredKeysAvailable(
        credentials,
        arrayOf(
          Pair("endpoint", "String"),
          Pair("metadata", "String"),
          Pair("captureNetworkQualityInPreview", "Boolean"),
        ),
      ) -> {
        config =
          HMSConfig(
            credentials.getString("username") as String,
            credentials.getString("authToken") as String,
            initEndpoint = credentials.getString("endpoint") as String,
            metadata = credentials.getString("metadata") as String,
            captureNetworkQualityInPreview =
              credentials.getBoolean("captureNetworkQualityInPreview"),
          )
      }
      areAllRequiredKeysAvailable(
        credentials,
        arrayOf(Pair("endpoint", "String"), Pair("metadata", "String")),
      ) -> {
        config =
          HMSConfig(
            credentials.getString("username") as String,
            credentials.getString("authToken") as String,
            initEndpoint = credentials.getString("endpoint") as String,
            metadata = credentials.getString("metadata") as String,
          )
      }
      areAllRequiredKeysAvailable(
        credentials,
        arrayOf(Pair("endpoint", "String"), Pair("captureNetworkQualityInPreview", "Boolean")),
      ) -> {
        config =
          HMSConfig(
            credentials.getString("username") as String,
            credentials.getString("authToken") as String,
            initEndpoint = credentials.getString("endpoint") as String,
            captureNetworkQualityInPreview =
              credentials.getBoolean("captureNetworkQualityInPreview"),
          )
      }
      areAllRequiredKeysAvailable(
        credentials,
        arrayOf(Pair("metadata", "String"), Pair("captureNetworkQualityInPreview", "Boolean")),
      ) -> {
        config =
          HMSConfig(
            credentials.getString("username") as String,
            credentials.getString("authToken") as String,
            metadata = credentials.getString("metadata") as String,
            captureNetworkQualityInPreview =
              credentials.getBoolean("captureNetworkQualityInPreview"),
          )
      }
      areAllRequiredKeysAvailable(credentials, arrayOf(Pair("endpoint", "String"))) -> {
        config =
          HMSConfig(
            credentials.getString("username") as String,
            credentials.getString("authToken") as String,
            initEndpoint = credentials.getString("endpoint") as String,
          )
      }
      areAllRequiredKeysAvailable(credentials, arrayOf(Pair("metadata", "String"))) -> {
        config =
          HMSConfig(
            credentials.getString("username") as String,
            credentials.getString("authToken") as String,
            metadata = credentials.getString("metadata") as String,
          )
      }
      areAllRequiredKeysAvailable(
        credentials,
        arrayOf(Pair("captureNetworkQualityInPreview", "Boolean")),
      ) -> {
        config =
          HMSConfig(
            credentials.getString("username") as String,
            credentials.getString("authToken") as String,
            captureNetworkQualityInPreview =
              credentials.getBoolean("captureNetworkQualityInPreview"),
          )
      }
    }
    return config
  }

  // TODO: replace surfaceView with hmsVideoView
  @RequiresApi(Build.VERSION_CODES.N)
  fun captureSurfaceView(
    surfaceView: SurfaceViewRenderer,
    sdkId: String,
    args: ReadableArray?,
    context: Context,
    id: Int,
  ) {
    val output = Arguments.createMap()
    if (args != null) {
      output.putInt("requestId", args.getInt(0))
    } else {
      output.putInt("requestId", -1)
    }
    val reactContext = context as ReactContext
    try {
      val bitmap: Bitmap =
        Bitmap.createBitmap(surfaceView.width, surfaceView.height, Bitmap.Config.ARGB_8888)
      PixelCopy.request(
        surfaceView,
        bitmap,
        { copyResult ->
          if (copyResult == PixelCopy.SUCCESS) {
            Log.d("captureSurfaceView", "bitmap: $bitmap")
            val byteArrayOutputStream = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream)
            val byteArray = byteArrayOutputStream.toByteArray()
            val encoded: String = Base64.encodeToString(byteArray, Base64.DEFAULT)
            Log.d("captureSurfaceView", "Base64: $encoded")
            output.putString("result", encoded)
            reactContext
              .getJSModule(RCTEventEmitter::class.java)
              .receiveEvent(id, "captureFrame", output)
          } else {
            Log.e("captureSurfaceView", "copyResult: $copyResult")
            HMSManager.hmsCollection[sdkId]?.emitHMSError(
              HMSException(
                103,
                copyResult.toString(),
                copyResult.toString(),
                copyResult.toString(),
                copyResult.toString(),
              ),
            )
            output.putString("error", copyResult.toString())
            reactContext
              .getJSModule(RCTEventEmitter::class.java)
              .receiveEvent(id, "captureFrame", output)
          }
        },
        Handler(),
      )
    } catch (e: Exception) {
      Log.e("captureSurfaceView", "error: $e")
      HMSManager.hmsCollection[sdkId]?.emitHMSError(e as HMSException)
      output.putString("error", e.message)
      reactContext.getJSModule(RCTEventEmitter::class.java).receiveEvent(id, "captureFrame", output)
    }
  }

  fun getAudioDevicesList(audioDevicesList: List<HMSAudioManager.AudioDevice>?): ReadableArray {
    val hmsAudioDevicesList = Arguments.createArray()
    if (audioDevicesList != null) {
      for (audioDevice in audioDevicesList) {
        hmsAudioDevicesList.pushString(audioDevice.name)
      }
    }
    return hmsAudioDevicesList
  }

  fun getAudioDevicesSet(audioDevicesSet: Set<HMSAudioManager.AudioDevice>?): ReadableArray {
    val hmsAudioDevicesSet = Arguments.createArray()
    if (audioDevicesSet != null) {
      for (audioDevice in audioDevicesSet) {
        hmsAudioDevicesSet.pushString(audioDevice.name)
      }
    }
    return hmsAudioDevicesSet
  }

  fun getAudioDevice(audioDevice: String?): HMSAudioManager.AudioDevice {
    when (audioDevice) {
      "AUTOMATIC" -> {
        return HMSAudioManager.AudioDevice.AUTOMATIC
      }
      "BLUETOOTH" -> {
        return HMSAudioManager.AudioDevice.BLUETOOTH
      }
      "EARPIECE" -> {
        return HMSAudioManager.AudioDevice.EARPIECE
      }
      "SPEAKER_PHONE" -> {
        return HMSAudioManager.AudioDevice.SPEAKER_PHONE
      }
      "WIRED_HEADSET" -> {
        return HMSAudioManager.AudioDevice.WIRED_HEADSET
      }
    }
    return HMSAudioManager.AudioDevice.SPEAKER_PHONE
  }

  fun getPeerListIteratorOptions(data: ReadableMap?): PeerListIteratorOptions? {
    if (data == null) {
      return null
    }
    val limit = data.getInt("limit")
    val role = data.getString("byRoleName")
    val peerIds = data.getArray("byPeerIds")?.toArrayList() as? ArrayList<String>

    return PeerListIteratorOptions(null, role, peerIds, limit)
  }
}
